KAFKA-19026: AlterConfigPolicy incompatibility between ZK mode and KR…

edoardocomar opened the pull request:
…aft mode when using AlterConfigOp.OpType.SUBTRACT

Modified ZkAdminManager.scala so that on OpType.SUBTRACT the policy receives the modified configs, as happens in KRaft mode.

This similarly fixes the OpType.APPEND differences.

Note that the policy behavior on OpType.DELETE is different when altering Broker and Topic resources. For topics the policy does not see a map entry, for brokers the config value is null. This was the existing behavior for KRaft and this commit does not change that.

ClusterTest added.

edoardocomar committed:
KAFKA-19026: AlterConfigPolicy incompatibility between ZK mode and KRaft mode when using AlterConfigOp.OpType.SUBTRACT

Modified ZkAdminManager.scala so that on OpType.SUBTRACT the policy
receives the modified configs, as happens in KRaft mode.

This similarly fixes the OpType.APPEND differences.

Note that the policy behavior on OpType.DELETE is different when altering Broker and Topic resources.
For topics the policy does not see a map entry, for brokers the config value is null.
This was the existing behavior for KRaft and this commit does not change that.

ClusterTest added.

edoardocomar commented:
Note that the inconsistent DELETE behavior between Topic and Broker resources is present in the KRaft mode and described by https://issues.apache.org/jira/browse/KAFKA-19028

edoardocomar commented:
@cmccabe @jsancio may you please take a look ? thanks

edoardocomar commented:
I'd love to have this fix in 3.9.1 ...

edoardocomar commented:
@showuon @frankvicky can we get this in 3.9.2 at least ?

edoardocomar commented:
@showuon would you like to give this PR a 2nd look ?

edoardocomar commented:
Hi i rebased on the current 3.9 branch and re-ran the test included in this PR
`./gradlew core:test --tests AlterConfigPolicyConfigsTest`

the test fails for ZK mode without the fix

edoardocomar commented:
IMHO, the failed tests in the CI Jenkins run are not related to this PR

FrankYang0529 commented:
Hi @edoardocomar, after discussing with @chia7712, we think that this is kind of behavior change in ZK mode. It’s not ideal to include such a change in a patch release like 3.9.2. I know we already added documentation about this in 3.9.1, which mentions the known issue https://issues.apache.org/jira/browse/KAFKA-19026. How about we enhance that document by adding a clear example and summarizing the conclusion we’ve reached here?

edoardocomar commented:
The reason I developed this fix - which I have applied to all our kafka clusters in ZK mode, and has been running for many months is that it allows to plug in a simple AlterConfigPolicy that disallows clearing a topic's `cleanup.policy`.
If a topic has an empty `cleanup.policy`, Kafka will allow it to grow forever. 

This in my opinion is an extremely important AlterConfigPolicy that the admin of a Kafka cluster should implement.
I don't think keeping backward compatibility in such a narrow field should prevail over correctness of behavior.

I am not interested to over document the issue. The release note was good enough for 3.9.1. Now it's a chance to get this fixed

FrankYang0529 commented:
@ijuma do you have any suggestion for this PR? ZK's AlterConfigPolicy behavior is long-standing. Do we really want to change it in 3.9.2 to match KRaft?

FrankYang0529 commented:
I create a discussion thread for this in dev channel.

https://lists.apache.org/thread/bknho1vrvby1tpq634g744vqf4mr8coh

edoardocomar committed:
KAFKA-19026: addition of KIP-1252 compatibility flag

With the introduction of KIP-1252, Kafka brokers running in
Zookeeper mode, can invoke an AlterConfigPolicy in a way that
is consistent with KRaft mode.

By default the flag is not enabled, so exixting behavior is preserved.

The unit test checks the behavior in KRaft mode and in Zookeeper mode
with both the flag enabled and disabled.

edoardocomar commented:
Hi all, as the KIP had 3 binding votes I could declare it accepted. However I preferred implementing it to see the end result.
I have rebased my old PR and added the flag and tests variations.
Please can anyone review ? Thank you

showuon reviewed (approved):
LGTM! Nice tests!

chia7712 commented on core/src/main/scala/kafka/server/ZkAdminManager.scala:
```diff
@@ -84,6 +83,8 @@ class ZkAdminManager(val config: KafkaConfig,
   private val alterConfigPolicy =
     Option(config.getConfiguredInstance(ALTER_CONFIG_POLICY_CLASS_NAME_CONFIG, classOf[AlterConfigPolicy]))
+  // KIP-1252 compatibility flag
+  private val alterConfigPolicyKraftCompatibilityEnabled = config.getBoolean(ALTER_CONFIG_POLICY_KRAFT_COMPATIBILITY_ENABLE_CONFIG)
```
would you mind adding this new configuration to `upgrade.md`?

edoardocomar replied:
done in https://github.com/apache/kafka/pull/21210

chia7712 commented on server-common/src/main/java/org/apache/kafka/server/config/ServerLogConfigs.java:
```diff
@@ -178,6 +178,12 @@ public class ServerLogConfigs {
     public static final String ALTER_CONFIG_POLICY_CLASS_NAME_CONFIG = "alter.config.policy.class.name";
     public static final String ALTER_CONFIG_POLICY_CLASS_NAME_DOC = "The alter configs policy class that should be used for validation. The class should " +
             "implement the <code>org.apache.kafka.server.policy.AlterConfigPolicy</code> interface.";
+    public static final String ALTER_CONFIG_POLICY_KRAFT_COMPATIBILITY_ENABLE_CONFIG = "alter.config.policy.kraft.compatibility.enable";
```
Keeping this configuration will cause an exception when upgrading to 4.x with the existing configuration file. We could make the 4.x broker ignore this configuration, or add a note to `upgrade.md` reminding users to remove it

edoardocomar replied:
I have added a note to the KIP - removable after migration,

edoardocomar replied:
please see https://github.com/apache/kafka/pull/21210

edoardocomar commented:
Hi @chia7712 I have a further PR for the release notes 
https://github.com/apache/kafka/pull/21210

The new config will be ignored by 4.x brokers if found in `server.properties` .
As with any custom config a user may add there - these get ignored but are retrievable at runtime from the maps passed to `Configurable` instances
